packages/yarnpkg-pnp/sources/loader/makeApi.ts

```diff
@@ -863,6 +880,28 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
 
   function resolveRequest(request: PortablePath, issuer: PortablePath | null, {considerBuiltins, extensions, conditions}: ResolveRequestOptions = {}): PortablePath | null {
     try {
+      if (request.startsWith(`#`)) {
+        assert(issuer, `An issuer is required to resolve private import mappings`);
+
+        const resolved = packageImportsResolve({
+          name: request,
+          base: pathToFileURL(npath.fromPortablePath(issuer)),
+          conditions: conditions ?? defaultExportsConditions,
+          readFileSyncFn: tryReadFile,
+        });
+
+        if (resolved instanceof URL) {
+          return resolveUnqualified(npath.toPortablePath(fileURLToPath(resolved)), {extensions});
+        } else {
+          if (resolved.startsWith(`#`))
+            // Node behaves interestingly by default so just block the request for now.
+            // https://github.com/nodejs/node/issues/40579
+            throw new Error(`Mapping from one private import to another isn't allowed`);
+
+          return resolveRequest(resolved as PortablePath, issuer, {conditions, considerBuiltins});
+        }
+      }
```

Comment:
Can you move this in a dedicated function (`resolvePrivateRequest`?)? No need to export it, I think, but it'd at least improve the readability.